jina/optimize/__init__.py

```diff
@@ -0,0 +1,231 @@
+import os
+from itertools import tee
+import json
+from pathlib import Path
+import shutil
+from ruamel.yaml import YAML
+
+from jina.flow import Flow
+from jina.helper import colored
+from jina.logging import default_logger as logger
+from type import IntegerParameter, load_optimization_parameters
+
+
+class FlowRunner:
+    def __init__(
+        self,
+        index_document_generator,
+        query_document_generator,
+        index_batch_size,
+        query_batch_size,
+        env_yaml=None,
+        workspace_env="JINA_WORKSPACE",
+        overwrite_workspace=False,
+    ):
+
+        self.index_document_generator = index_document_generator
```

Comment:
why do we have generators and batch_sizes as members of the class and not passed to the `run` methods. Remember generators can not be reused for iteration again.

Reply:
`tee` is used to make a generator copy and rerun generators